scrapy/http/headers.py

```diff
@@ -76,13 +77,13 @@ def to_string(self):
         return headers_dict_to_raw(self)
 
     def to_unicode_dict(self):
-        """ Return headers as a CaselessDict with unicode keys
+        """ Return headers as a CaseInsensitiveDict with unicode keys
         and unicode values. Multiple values are joined with ','.
         """
-        return CaselessDict(
-            (to_unicode(key, encoding=self.encoding),
-             to_unicode(b','.join(value), encoding=self.encoding))
-            for key, value in self.items())
+        return CaseInsensitiveDict(
+            (to_unicode(key, encoding=self.encoding), to_unicode(b','.join(value), encoding=self.encoding))
+            for key, value in self.items()
+        )
```

Comment:
This change should be safe I think, `CaseInsensitiveDict`'s interface is the same as `CaselessDict`. The method is only used [here](https://github.com/scrapy/scrapy/blob/2.5.0/scrapy/core/downloader/handlers/s3.py#L67).

Reply:
It’s still a potential issue for `isinstance` usages. I’m not saying I’m against this change, but I am unsure.

Reply:
This should address the `isinstance` check:
```diff
diff --git scrapy/utils/datatypes.py scrapy/utils/datatypes.py
index 6eeabe1e..63c07b48 100644
--- scrapy/utils/datatypes.py
+++ scrapy/utils/datatypes.py
@@ -21,7 +21,7 @@ class CaselessDict(dict):
     def __new__(cls, *args, **kwargs):
         from scrapy.http.headers import Headers
 
-        if issubclass(cls, CaselessDict) and not issubclass(cls, Headers):
+        if issubclass(cls, CaselessDict) and not issubclass(cls, (CaseInsensitiveDict, Headers)):
             warnings.warn(
                 "scrapy.utils.datatypes.CaselessDict is deprecated,"
                 " please use scrapy.utils.datatypes.CaseInsensitiveDict instead",
@@ -79,7 +79,7 @@ class CaselessDict(dict):
         return dict.pop(self, self.normkey(key), *args)
 
 
-class CaseInsensitiveDict(collections.UserDict):
+class CaseInsensitiveDict(collections.UserDict, CaselessDict):
     """A dict-like structure that accepts strings or bytes as keys and allows case-insensitive lookups.
 
     It also allows overriding key and value normalization by defining custom `normkey` and `normvalue` methods.
```

Although as we discussed this might not be a problem, mostly because `CaselessDict` is not documented.